rfcs/0174-improve-terminology.md

```diff
@@ -0,0 +1,160 @@
+
+- Nix expression doesn't necessarily contain a derivation. Using `nix eval` to evaluate it will produce a result without producing a .drv file
+- Nix can be configured to not keep .drv files in the store using a setting `keep-derivations` (default is `true`)
+- Fixed-Output Derivations, or FODs, are mostly not about derivations. They are about a type of store path that has a well-known hash that is compared with the actual hash
+- Import From Derivation, or IFD, isn't about derivations, either! It is about reading data from a store path
+- Packages that exist in Nixpkgs are often called "derivations", despite the existence of a buildspec in .drv file being an implementation detail
+- Nix store outputs that aren't exposed in Nixpkgs are called "derivations". Examples include store paths passed to `src`, symlink farms and other kinds of store paths used in implementation, but not exposed to the end user
+
+In summary, "derivation" concept is an implementation detail. The term, however, is applied to many user-facing scenarios. As a result, the term becomes overloaded, confusing and difficult to comprehend. We should treat derivations as an implementation detail.
+
+Continuing on the topic of derivations, it is clear that store path outputs aren't always directly produced by a derivation. It is not very useful to think in "derivation -> output" chains. Instead, we can use a term "installable" that is [already established](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix#installables) to simplify the terminology:
+
+- `nix build` is a command that converts an installable to an output store path or fails. It is a simple "input -> output" relationship
+- In this model, what is written to the store (aside from the output) are just *dependencies*. This includes plain dependencies (like software libraries), FODs, implementation-based store paths and .drv files
+- Installables are usually packages or runnable scripts. That makes them easy to conceptualize, and might give additional intuition into the future Flake transition. The current unstable schema for Flakes uses installables for all outputs. That said, Flakes also allow definining custom outputs that can be arbitrary
+
+With all that reasoning, here are the actions proposed:
+
+- Rename IFD (Import From Derivation) to IFSP (Import From Store Path)
+- Rename FOD (Fixed-Output Derivation) to WKSP (Well-Known Store Path)
+- Converge on using "package" term for introductory materials
+- Converge on using "installable" term for more advanced materials. This includes cases when an installable is a runnable script, such as when building a closure of NixOS system. The term should also be used in manuals and references
+- Rename "derivation" to "derivation file", keeping the current definition. Define "derivation" as "an old term that is no longer used, which usually means derivation files"
```

Comment:
But this is incorrect... That isn't the current definition. `*.drv` files are [store derivations,](https://nixos.org/manual/nix/stable/glossary#gloss-store-derivation) which are distinct from derivations. Derivations are specifications for "running an executable on precisely defined input files to repeatably produce output files at uniquely determined file system paths."[^1] Notably, store derivations are a side-effect of derivations.

[^1]: https://nixos.org/manual/nix/stable/language/derivations

Reply:
Looking at the glossary definition from Nix manual for [derivation](https://nixos.org/manual/nix/stable/glossary#gloss-derivation), it reads as if derivation is just a structure. This is some structure that describes a build task, and Nix package managers operates with it. Crucially, the structure can be in any format at all: it can be passed from a C++ program, using the code that Nix package manager uses; it can be a JSON; it can be an aTerm; it can also be some area in RAM that holds a representation of the C++ structure in memory.

This definition isn't useful. It is essentially saying, "there's a structure that is defined in C++ code, and everything that can be serialized into it is a derivation". This is not even an implementation detail, this is more like misplaced code comment.

Notably, the definition for [store derivation](https://nixos.org/manual/nix/stable/glossary#gloss-store-derivation) that you linked is much more useful. It outright states a couple of useful properties: store derivation is a derivation, it exists in store and has a store path, and it is also a .drv file. Since store derivation is actually a derivation, and it's also the predominant format of interacting with derivations even if you interface with the Nix package manager code - I'd say it makes a lot of sense to stop trying to describe what's essentially an internal structure representation, and to only describe the .drv file.

Also, that's only going by the glossary. The less thorough parts of the documentation often use "derivation" to mean "store derivation". And less formal knowledge sources pretty much always say "derivation" to mean the .drv file.

It's important to note that all of the above is not what is actually written in the text of the RFC, so that's a mistake on my part that I'll fix.

Reply:
A derivation indeed is just an abstract data type, the implementation doesn't really matter. And yes, many places in the Nix manual are still sloppy on the terms.

I generally agree on the premise of this RFC, but strongly doubt it's productive to pursue: batching up terms for renaming will just delay progress on the uncontroversial ones, and clutter the discussion for the others.

Some of the terms you've listed have issues filed for finding better names (IFD). Some are arguably already solved but not used consistently yet (Nix language), some are merely not well-explained (derivation), some can be considered to be not problematic to begin with. All of that can be improved incrementally, and I'd very much appreciate your help with that.


Some terms probably aren't worth changing: Seriously trying to rename Nix is asking for trouble.

In any case I recommend to take on whatever you feel needs to be done as independent issues.